test/assignment-7/AMM.t.sol

```diff
@@ -0,0 +1,194 @@
+
+        assertEq(z, 50 * 1e36);
+        assertEq(amm.balanceOf(alice), 50 * 1e36);
+    }
+}
+
+abstract contract InitializedState is ZeroState {
+    function setUp() public override virtual {
+        super.setUp();
+
+        token0.mint(alice, 99999 ether);
+        token1.mint(alice, 99999 ether);
+
+        vm.prank(alice);
+        amm.init(5 ether, 10 ether);
+    }
+}
+
+contract InitializedStateTest is InitializedState {
+
+    event Mint(address indexed user, uint256 amount0, uint256 amount1, uint256 amountLP);
+    event Burn(address indexed user, uint256 amount0, uint256 amount1, uint256 amountLP);
+    event Sell0(address indexed user, uint256 amount0, uint256 amount1);
+    event Sell1(address indexed user, uint256 amount0, uint256 amount1);
+
+    function testRevertsIfCallingInitTwice() public {
+        token0.mint(alice, 1 ether);
+        token1.mint(alice, 1 ether);
+
+        vm.prank(alice);
+        vm.expectRevert(AMM.AlreadyInitialized.selector);
+        amm.init(1 ether, 1 ether);
+    }
+
+    function testMintSameAmountShouldMintSameAmountOfLPTokens() public {
+        vm.prank(alice);
+        vm.expectEmit(true, true, true, true);
+        emit Mint(alice, 5 ether, 10 ether, 50 * 1e36);
+        uint z = amm.mint(5 ether, 10 ether);
+
+        assertEq(z, 50 * 1e36);
+        assertEq(amm.balanceOf(alice), 100 * 1e36);
+    }
+
+    function testMintDifferentAmount() public {
+        vm.prank(alice);
+        uint z = amm.mint(0.1234 ether, 0.2468 ether);
+
+        assertEq(z, (0.1234/5) * 50 * 1e36);
+    }
+
+    function testMintRevertsIfWrongProportion() public {
+        vm.prank(alice);
+        vm.expectRevert(AMM.IncorrectProportion.selector);
+        amm.mint(5 ether, 9 ether);
+    }
+
+    function testBurn() public {
+        // Alice current LP token balance
+        assertEq(amm.balanceOf(alice), 50 * 1e36);
+
+        // Burn all alice tokens for easy calculation
+        token0.burn(alice, token0.balanceOf(alice));
+        token1.burn(alice, token1.balanceOf(alice));
+
+        vm.prank(alice);
+        vm.expectEmit(true, true, true, true);
+        emit Burn(alice, 1 ether, 2 ether, 10 * 1e36);
+        (uint x, uint y) = amm.burn(10 * 1e36);
+
+        assertEq(x, 1 ether);
+        assertEq(y, 2 ether);
+
+        assertEq(token0.balanceOf(alice), 1 ether);
+        assertEq(token1.balanceOf(alice), 2 ether);
+    }
+
+    function testSell0() public {
+        token0.mint(bob, 10 ether);
+        uint256 k = amm.reserve0() * amm.reserve1();
+
+        // expected swap amount: (1*10)/(5+1) = 1.6666...
+        vm.prank(bob);
+        vm.expectEmit(true, true, true, true);
+        emit Sell0(bob, 1 ether, uint(5) * 1e18 / 3);
+        uint amount1 = amm.sell0(1 ether);
+        
+        assertEq(amount1, uint(5) * 1e18 / 3);
+        assertEq(token1.balanceOf(bob), amount1);
+
+        // constant product is maintained
+        assertRelApproxEq(amm.reserve0() * amm.reserve1(), k, 0);
```

Comment:
The constant product is what we call an invariant, and a great target for fuzzing tests.

Reply:
👍 
added a fuzz test here: https://github.com/davidbrai/yield-mentorship/pull/8/commits/a18d3b9b69b420501c1f1c050ca7f9d4cb8fa4eb

it's actually a good way to force you to think about what happens at the edges, i.e maximum/minimum values.
for now I just limited the range of the input, but if I were building this AMM more seriously then it would require more thinking

Reply:
That's exactly it.